frontend/src/lib/v2/CompareUtils.ts

```diff
@@ -55,6 +56,95 @@ export interface RunArtifactData {
   artifactCount: number;
 }
 
+export interface NameId {
+  name: string;
+  id: string;
+}
+
+// The provided name takes the format of "<type> ID #<number>" if the display name is not available.
+export interface FullArtifactPath {
+  run: NameId;
+  execution: NameId;
+  artifact: NameId;
+}
+
+export interface RocCurveArtifactData {
+  validLinkedArtifacts: LinkedArtifact[];
+  fullArtifactPathMap: { [key: string]: FullArtifactPath };
+}
+
+export const getRocCurveId = (linkedArtifact: LinkedArtifact): string =>
+  `${linkedArtifact.event.getExecutionId()}-${linkedArtifact.event.getArtifactId()}`;
+
+// Form an array which holds all valid ROC Curve linked artifacts.
+export const getValidRocCurveLinkedArtifacts = (
+  rocCurveRunArtifacts: RunArtifact[],
+): RocCurveArtifactData => {
+  const fullArtifactPathMap: { [key: string]: FullArtifactPath } = {};
+  const validLinkedArtifacts = flatMapDeep(
+    rocCurveRunArtifacts.map(runArtifact =>
+      runArtifact.executionArtifacts.map(executionArtifact => {
+        return getValidArtifacts(runArtifact, executionArtifact, fullArtifactPathMap);
```

Comment:
I've updated the logic so this present function only returns the valid linked artifacts, and the modification of the `fullArtifactPathMap` is generated by looping through the returned function's value: https://github.com/kubeflow/pipelines/pull/8073/commits/336df9615ef654e641d461ad115506a26292cb36.

Is this process clearer / more explicit? This does add to the runtime a little bit because it loops through the linked artifacts twice, but it does not change the big-O runtime.

Reply:
Sounds good! The code looks good with the update. Although we want to save execution cost, we also need to consider the readability. For example: We can refactor the function so that `mutation to fullArtifactPathMap` happens in the same logic block that `fullArtifactPathMap` is defined, instead of being wrapped inside `getValidArtifacts()`. This is non-blocking so we can discuss it later.

Reply:
Yes agreed - I did in fact move the `mutation to fullArtifactPathMap` from within `getValidArtifacts()` to where it is defined, instead using the return result from `getValidArtifacts()`. Discussing later sounds good as well 👍